src/sources/dnstap/dns_message_parser.rs

```diff
@@ -0,0 +1,2000 @@
+use super::dns_message::{
+    self, DnsQueryMessage, DnsRecord, DnsUpdateMessage, EdnsOptionEntry, OptPseudoSection,
+    QueryHeader, QueryQuestion, UpdateHeader, ZoneInfo,
+};
+use data_encoding::{BASE32HEX_NOPAD, HEXUPPER};
+use snafu::{ResultExt, Snafu};
+use std::str::Utf8Error;
+use trust_dns_proto::{
+    error::ProtoError,
+    op::{message::Message as TrustDnsMessage, Edns, Query},
+    rr::{
+        dnssec::{rdata::DNSSECRData, Algorithm, SupportedAlgorithms},
+        rdata::{
+            a, aaaa,
+            caa::Value,
+            opt::{EdnsCode, EdnsOption},
+            NULL,
+        },
+        record_data::RData,
+        resource::Record,
+        Name, RecordType,
+    },
+    serialize::binary::{BinDecodable, BinDecoder},
+};
+
+#[derive(Debug, Snafu)]
+pub enum DnsMessageParserError {
+    #[snafu(display("Encountered error : {}", cause))]
+    SimpleError { cause: String },
+    #[snafu(display("Encountered error from TrustDns: {}", source.to_string()))]
+    TrustDnsError { source: ProtoError },
+    #[snafu(display("UTF8Error: {}", source))]
+    Utf8ParsingError { source: Utf8Error },
+}
+
+pub struct DnsMessageParser {
+    raw_message: Vec<u8>,
+    raw_message_clone: Option<Vec<u8>>, // Used only for parsing compressed domain names
```

Comment:
Could you explain a bit more what this field is used for? I'm wondering if we could give it a better name.

Reply:
As the comment said, `raw_message_clone` is used to facilitate parsing of compressed domain names contained in a record data (i.e. `rdata`). It's basically a copy of the `raw_message`, but with a doubled capacity to accommodate all `rdata` that may contain compressed domain name.
We could instantiate a new  copy of the `raw_message` whenever we need to parse `rdata`, but it'd obviously impact performance more or less. That's why we ended up the current solution: upon the first call to parse an `rdata` that may contain compressed domain name, we'll instantiate a copy of the `raw_message` with double-sized capacity, and store it in the struct member field `raw_message_clone`; for subsequent invocations on the same call, we simply reuse the already-instantiated member field `raw_message_clone`.
And yes, I totally agree that the member field's current name, `raw_message_clone`, is both ambiguous and meaningless.I guess `raw_message_for_rdata_parsing` could be a better name?

Reply:
Updated.